### ISSUE TITLE
[STORY] Create analysis class decorators

### ISSUE THREAD

**Issue by tclose** [Australian-Imaging-Service/arcana#9]
### Description

Reimplement Analysis classes in v2.0 using `attrs` decorators to create our own DSL, e.g.

```python
from pydra.tasks.fsl.preprocess.bet import Bet
from arcana.mark import analysis, data_column, parameter, pipeline
from arcana.data.formats.medimage import MedImage, Dicom, Nifti, NiftiXGz
from arcana.salience import DataSalience as ds, ParamSalience as ps
from arcana.data.dimensions.medimage import ClinicalTrial as per


@analysis
class T1w():

    # Data columns
    magnitude: MedImage = data_column(salience=ds.primary, desc="The primary magnitude image, typically reconstructed on the MRI scanner")
    brain: NiftiXGz = data_column(salience=ds.qa, desc="The brain extracted image")
    brain_mask: NiftiXGz = data_column(salience=ds.qa, desc="The brain mask used for the analysis")
    grey_matter_thickness: float = data_column(salience=ds.publication, desc="The average grey matter thickness across a subject's brain")
    average_gm_thickness: float = data_column(salience=ds.publication, frequency=per.dataset, desc="Average GM thickness averaged across dataset")

    # Parameters
    bet_method: str = parameter(salience=ps.recommended, default='robust', options=('robust', 'bias_and_neck'), help="The method used to extract the brain mask")
    bet_threshold: float = parameter(salience=ps.arbitrary, default=0.5, help="The threshold at which the mask is applied")

    @pipeline(
        brain,
        brain_mask)
    def brain_mask(self, pipeline, magnitude: Nifti, bet_method: str, bet_threshold: float):  # column and parameters are automagically passed to function when it is called

        pipeline.add(BET(
            name='mask_task',
            in_file=magnitude.path,
            method=bet_method,
            threshold=bet_threshold)

       return pipeline.mask_task.lzout.brain, pipeline.mask_task.lzout.mask
```

### Acceptance Criteria

- [ ] 1. Analysis classes implemented such that they match description/examples in arcana.readthedocs.io documentation
- [ ] 2. Newly added code is 90% covered by unittests
- [ ] 3. Classes are validated to ensure pipelines are properly linked to columns and parameters

**Comment by tclose** [comment]
A multi-contrast analysis could look something like

```python
from arcana.mark import analysis, data_column, parameter, pipeline, subanalysis
from arcana.salience import DataSalience as ds, ParamSalience as ps
from arcana.analyses.bids.mri.anat import T1w, T2w, Bold, Dwi

@analysis
class BidsMri():

    # Data columns
    combined_connectivity: Matrix = data_column(salience=ds.publication, desc="The combined structural/functional connectivity matrix"

    # Parameters
    connectivity_weighting: float = parameter(salience=ps.arbitrary, desc="The weighting between function and structural connectivity in combined matrix")

    # Sub-analyses
    t2w = subanalysis(T2w)
    t1w = subanalysis(T1w, peel_contrast=t2w.brain)
    bold = subanalysis(Bold, white_grey_surface=t1w.white_grey_surface)
    dwi = subanalysis(Dwi, white_grey_surface=t1w.white_grey_surface)


    @pipeline(
        'combined_connectivity',
        structural_matrix=dwi.connectivity_matrix,
        functional_matrix=bold.connectivity_matrix)
    def combined_connectivities(self, pipeline, structural_matrix: Matrix, functional_matrix: Matrix, connectivity_weighting: float):

        pipeline.add(CombineConnectivity(
            name='combine',
            fmri_matrix=fmri_matrix,
            dwi_matrix=dwi_matrix,
            weighting=dwi_connectivity_matrix)

       return pipeline.combine.lzout.out_file
  
```

**Comment by tclose** [comment]
If you want to customise an analysis and use an alternative tool it would look something like

```python
from arcana.types import Kspace
from pydra.tasks.recon import KspaceRecon


@analysis
class RawT1w(T1w):

    # Data columns
    kspace: Kspace = data_column(salience=ds.primary, help="T1w image in k-space format instead of reconstructed DICOM/NIfTI")

    @pipeline(
        'magnitude',
        condition=kspace.provided and not T1w.magnitude.provided)
    def recon(self, pipeline, kspace: Nifti):

        pipeline.add(KspaceRecon(
            name='recon',
            in_file=kspace.path)

       return pipeline.recon.out_file
```

or if you wanted to

**Comment by tclose** [comment]
or if you wanted to add the ability to reconstruct from k-space

```python
from pydra.tasks.recon import KspaceRecon
from arcana.formats.biomedimage import Dicom
from arcana.types.mri.raw import Kspace, Ismrmrd
from arcana.analyses.bids.anat import T1w


@analysis
class RawT1w(T1w):

    # Data columns
    kspace: Kspace = data_column(salience=ds.primary, help="T1w image in k-space format instead of reconstructed DICOM/NIfTI")

    @pipeline(
        ('magnitude', Dicom),  # Implicit conversion from produced DICOM format to that required by downstream pipelines will be performed
        condition=kspace.provided and not T1w.magnitude.provided)
    def recon(self, pipeline, kspace: Ismrmrd):  # Implicit conversion from provided kspace format to ISMRM-RD format will be performed by Arcana

        pipeline.add(KspaceRecon(
            name='recon',
            in_file=kspace.path)

       return pipeline.recon.lzout.out_file
```

**Comment by tclose** [comment]
Epic: Australian-Imaging-Service/pipelines#17

**Comment by tclose** [comment]
Implemented in #125
